### ISSUE TITLE
Losses + OneHotArray support

### ISSUE THREAD

**Issue by darsnack** [FluxML/Flux.jl#1493]
Opening this after #1492. While we don't necessarily want the specific changes in that PR, I think we should look through the loss functions and identify some fast paths when the true labels, `y isa OneHotArray`. Probably many of them can be generalized into updates to `onehot.jl`, though some might require individual dispatch.

I'll come back to this later to update with a list of potential paths (if any). Someone else is also welcome to edit if they get to it before me.

**Comment by findmyway** [comment]
By the way, the CUDA support also needs to be considered.

For example, currently the following code will trigger the warning:

```julia
julia> cu(Flux.OneHotMatrix([3,2,3], 3))
3×3 Flux.OneHotArray{Int64,3,1,2,CuArray{Int64,1}}:
┌ Warning: Performing scalar operations on GPU arrays: This is very slow, consider disallowing these operations with `allowscalar(false)`
└ @ GPUArrays ~/.julia/packages/GPUArrays/WV76E/src/host/indexing.jl:43
 0  0  0
 0  1  0
 1  0  1
```

**Comment by darsnack** [comment]
That shouldn't be happening! Let me look into it, and can you file a separate issue for this?

**Comment by findmyway** [comment]
Filed at https://github.com/FluxML/Flux.jl/issues/1494
